### ISSUE TITLE
Does it support IOS and Android?

### ISSUE THREAD

**Issue by Personuo** [psydack/uimgui#50]
I want to use it in my project immediately, but I don't know if ios and android are supported?

**Comment by psydack** [comment]
Hey! I never tested before. Please, if you can test it, let me know if it
works.

Thank you for your suggestion.
Be in mind that ImGui has some issues in your core related to mobile

**Comment by cholushkin** [comment]
I've just tested on android. It doesn't render anything. Later I'll check the logs

**Comment by cholushkin** [comment]
So the error is in loading cimgui.dll on android: 

2023/08/11 16:37:47.683 18317 18338 Error Unity DllNotFoundException: Unable to load DLL 'cimgui'. Tried the load the following dynamic libraries: Unable to load dynamic library 'cimgui' because of 'Failed to open the requested dynamic library (0x06000000) dlerror() = dlopen failed: library "cimgui" not found
2023/08/11 16:37:47.683 18317 18338 Error Unity   at ImGuiNET.ImGuiNative.igCreateContext (ImGuiNET.ImFontAtlas* shared_font_atlas) [0x00000] in <00000000000000000000000000000000>:0 
2023/08/11 16:37:47.683 18317 18338 Error Unity   at UImGui.UImGuiUtility.CreateContext () [0x00000] in <00000000000000000000000000000000>:0 
2023/08/11 16:37:47.683 18317 18338 Error Unity   at UImGui.UImGui.Awake () [0x00000] in <00000000000000000000000000000000>:0

**Comment by nukadelic** [comment]
Same here works flawlessly inside the editor but won't render any uimgui once build ( app runs fine on its own ) tried including the .so with x86 & armv7 file from [ImGui.NET-nativebuild/releases](https://github.com/ImGuiNET/ImGui.NET-nativebuild/releases) ( although I suspect the versions are way off ) , tried both mono and cpp build but still no luck , not very versed with native plugins , maybe someone could figure it out here are a few links I came across : [DllNotFoundException on Android](https://stackoverflow.com/questions/44027714/dllnotfoundexception-on-android) , [Create a .jar from a .dll](https://stackoverflow.com/questions/28155257/create-a-jar-from-a-dll)

**Comment by nukadelic** [comment]
Also [tried building](https://github.com/ImGuiNET/ImGui.NET-nativebuild/issues/32) my own `.so` file without any luck so far

**Comment by noncom** [comment]
Disclaimer: this is a guess.

I read the description of what you did by the link in your last comment, and I think it won't work like that in Unity.
The problem is likely that the code that requires the dll/so is inside the already built ImGui.NET.dll.
I suspect that in order for this to work, all the `[DLLImport]` statements should be resolved and compiled directly by Unity itself.

If this guess is correct then you might probably have success with:
- Copying the source form of the ImGui.NET, including the generated code into the Unity project
- Copying the source code of uimgui (this project) into your Unity project
- Reorganizing the positioning of cimgui according to the default Unity pattern of multiplatform native libraries (see the article cited below)
- Maybe doing something similar to all the other dlls used by uimgui

In that case the whole project, including the linking information for the native libraries is going to be produced and managed by Unity, which should allow it to correctly include the libraries in the build, and be able to resolve them in runtime.

A brief article with generic steps, just for reference: https://matiaslavik.wordpress.com/2021/01/21/cross-platform-native-plugins-in-unity/

**Comment by nukadelic** [comment]
> Disclaimer: this is a guess.
> 
> I read the description of what you did by the link in your last comment, and I think it won't work like that in Unity. The problem is likely that the code that requires the dll/so is inside the already built ImGui.NET.dll. I suspect that in order for this to work, all the `[DllImport]` statements should be resolved and compiled directly by Unity itself.
> 
> If this guess is correct then you might probably have success with:
> 
>     * Copying the source form of the ImGui.NET, including the generated code into the Unity project
> 
>     * Copying the source code of uimgui (this project) into your Unity project
> 
>     * Reorganizing the positioning of cimgui according to the default Unity pattern of multiplatform native libraries (see the article referenced below)
> 
>     * Maybe doing something similar to all the other **native** dlls used by ImGui.NET and uimgui
> 
> 
> In that case the whole project, including the linking information for the native libraries is going to be produced and managed by Unity, which should allow it to correctly include the libraries in the build, and be able to resolve them in runtime.
> 
> A brief article with generic steps, just for reference: https://matiaslavik.wordpress.com/2021/01/21/cross-platform-native-plugins-in-unity/

Can you dm me on discord ? same handle as my github username

**Comment by rbetik12** [comment]
Hi everyone! I tried to compile cimgui for android arm32/64 as a separate .so file and it worked out of the box.

First of all I've compiled the cimgui for a correct version of imgui used in this repo 1.90.1:

1. Cloned [ImGuiNet-nativebuild](https://github.com/ImGuiNET/ImGui.NET-nativebuild/releases/tag/v1.90.1) for release 1.90.1
2. Opened cimgui submodule in my terminal (I am using Win11)
3. Checked that I've android sdk and ndk installed
4. Compiled cimgui .so for arm32 and arm64 with that command (run that in cimgui/build):
```
cmake .. -DANDROID_ABI=<armeabi-v7a or arm64-v8a> -DANDROID_PLATFORM=22 (you can use whatever SDK version you want) -DANDROID_NDK="<Android SDK path>\ndk\27.1.12297006" -DCMAKE_TOOLCHAIN_FILE="<Android SDK path>\ndk\27.1.12297006\build\cmake\android.toolchain.cmake" -G Ninja
```
5. After that I run another cmake command to compile the .so
`cmake --build build --config Release --parallel <cpu count>`
6. After that I renamed `cimgui.so -> libcimgui.so` and put it to my Unity project
7. In Unity I've set that it must be used only on Android platform and set the correct architecture
![image](https://github.com/user-attachments/assets/ef47741b-715d-419c-aea3-e87a5d48c41f)
8. After that I've launched my game and android and got the ImGui window, but it renders with incorrect colors, beside that everything works as good as on desktop
![image](https://github.com/user-attachments/assets/a2829284-be82-4a5a-8af4-35b90e9dd926)
9. _Just a small note_ I've configured font size and imgui scale for window to be usable size on the device
```
            var io = ImGui.GetIO();
            var style = ImGui.GetStyle();
            float defaultScale = 1.0f;

            if (Application.platform == RuntimePlatform.Android || Application.platform == RuntimePlatform.IPhonePlayer)
            {
                defaultScale = 2.5f;
            }
            
            Debug.Log($"[CheatUIManager] Default imgui scale: {defaultScale}");

            io.DisplaySize = new Vector2(defaultScale, defaultScale);
            style.FrameRounding = 0.0f;
            style.FrameBorderSize = 0.5f;
            style.ScaleAllSizes(defaultScale);
            
            var font = io.Fonts.Fonts[0];
            font.Scale = defaultScale;

```

@psydack Do you have any info on how to fix that incorrect colors on android (I am using only OpenGL unity backend) and it seems their's a problem with the transparency

**Comment by nukadelic** [comment]
can you share the .so file ?

**Comment by rbetik12** [comment]
[libs.zip](https://github.com/user-attachments/files/17318059/libs.zip)

Here you go, arm32 and arm64 versions attached

**Comment by nickleme** [comment]
I've tested this in my project and it's pretty much plug and play! The colors are indeed a bit off but it's working. 

Thank you @rbetik12 for providing the .so files!

**Comment by rbetik12** [comment]
Now we need someone to test that on iOS

**Comment by nukadelic** [comment]
apk on the Quest-2 crashes , no unity errors / logcat doesn't show any errors - no clue how to check what's going on

**Comment by rbetik12** [comment]
I guess you can try compiling .so files by yourself, using specific SDK Quest-2 uses

**Comment by nukadelic** [comment]
yeah already did try before a few times ( without any luck ) , thanks for sharing the `.so` files tho 

Would you mind sharing a your build process - step by step including all the tools and flags you have used ? Curious to try your method - because your `.so` file just crashes , whenever i try to activate the imgui . When i was trying to build with my method it wouldn't even load the DLL ( `DllNotFoundException: Unable to load DLL 'cimgui'.` ) - so its a good sign that it successfully loads , but might be missing something.

**Comment by rbetik12** [comment]
Have you tried disabling in Project Settings > Player > Other Settings > Script define symbols > Apply > Restart Unity Editor.

UIMGUI_REMOVE_IMPLOT
UIMGUI_REMOVE_IMNODES
UIMGUI_REMOVE_IMGUIZMO

Also do you have any callstacks related to crash?

**Comment by nukadelic** [comment]
That's the odd part , there aren't any errors , tried building in dev mode with untiy debugger , tried watching logcat , it simply quits as soon as i enable the imgui panel ( which seems to work fine in editor ) 

I'll try building with those define symbols , hopefully something will show up

**Comment by tonycoculuzzi** [comment]
Has this been solved and implemented?

**Comment by ryanw3bb** [comment]
> Hi everyone! I tried to compile cimgui for android arm32/64 as a separate .so file and it worked out of the box.
> 
> First of all I've compiled the cimgui for a correct version of imgui used in this repo 1.90.1:
> 
> 1. Cloned [ImGuiNet-nativebuild](https://github.com/ImGuiNET/ImGui.NET-nativebuild/releases/tag/v1.90.1) for release 1.90.1
> 2. Opened cimgui submodule in my terminal (I am using Win11)
> 3. Checked that I've android sdk and ndk installed
> 4. Compiled cimgui .so for arm32 and arm64 with that command (run that in cimgui/build):
> 
> ```
> cmake .. -DANDROID_ABI=<armeabi-v7a or arm64-v8a> -DANDROID_PLATFORM=22 (you can use whatever SDK version you want) -DANDROID_NDK="<Android SDK path>\ndk\27.1.12297006" -DCMAKE_TOOLCHAIN_FILE="<Android SDK path>\ndk\27.1.12297006\build\cmake\android.toolchain.cmake" -G Ninja
> ```
> 
> 5. After that I run another cmake command to compile the .so
>    `cmake --build build --config Release --parallel <cpu count>`
> 6. After that I renamed `cimgui.so -> libcimgui.so` and put it to my Unity project
> 7. In Unity I've set that it must be used only on Android platform and set the correct architecture
>    ![image](https://github.com/user-attachments/assets/ef47741b-715d-419c-aea3-e87a5d48c41f)
> 8. After that I've launched my game and android and got the ImGui window, but it renders with incorrect colors, beside that everything works as good as on desktop
>    ![image](https://github.com/user-attachments/assets/a2829284-be82-4a5a-8af4-35b90e9dd926)
> 9. _Just a small note_ I've configured font size and imgui scale for window to be usable size on the device
> 
> ```
>             var io = ImGui.GetIO();
>             var style = ImGui.GetStyle();
>             float defaultScale = 1.0f;
> 
>             if (Application.platform == RuntimePlatform.Android || Application.platform == RuntimePlatform.IPhonePlayer)
>             {
>                 defaultScale = 2.5f;
>             }
>             
>             Debug.Log($"[CheatUIManager] Default imgui scale: {defaultScale}");
> 
>             io.DisplaySize = new Vector2(defaultScale, defaultScale);
>             style.FrameRounding = 0.0f;
>             style.FrameBorderSize = 0.5f;
>             style.ScaleAllSizes(defaultScale);
>             
>             var font = io.Fonts.Fonts[0];
>             font.Scale = defaultScale;
> ```
> 
> [@psydack](https://github.com/psydack) Do you have any info on how to fix that incorrect colors on android (I am using only OpenGL unity backend) and it seems their's a problem with the transparency

Hey @rbetik12 did you ever work out why the colors are incorrect on Android? I tried reordering the unpack_color() bitshift in the shader which improved things, but they still don't look correct. Debugging a little further shows the vertex colors coming in from ImGui are incorrect on Android compared to running in Editor, which is odd. @psydack any ideas?

**Comment by nukadelic** [comment]
Yes -> [Direct Link to .SO file](https://github.com/yCatDev/uimgui-extended/blob/main/Plugins/imgui/android/arm64-v8a/libcimgui.so)

In my case to get it working on the Oculus Quest ( for an apk build ) seems like it has to be build on Linux , someone else has successfully build an .so that actually loads on the quest and seems to get imgui working on the Quest ->> [Discussion link , he posted his repo at the end ](https://github.com/ImGuiNET/ImGui.NET-nativebuild/issues/32)
